### PR TITLE
SDEV-986  Zip SAST file and send it to IQ

### PR DESCRIPTION
**Pull request by oospinar**
Send SAST file to IQ server as part of the request.

### Testing done

Test done on local environment, run Jenkins with the installed plugin, create e new project, add the sast.json file as part of the pipeline then validates that the iq process the sast.json file.

Related links:

* PRs: 
https://github.com/sonatype/nexus-java-api/pull/278
https://github.com/sonatype/insight-scanner/pull/392
https://github.com/sonatype/insight-brain/pull/11343

* Jira
https://sonatype.atlassian.net/browse/SDEV-986

```[tasklist]
### Submitter checklist
- [X] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or Jira
- [X] Link to relevant pull requests, esp. upstream and downstream changes
- [X] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -89,7 +89,7 @@
     <jvnet-localizer-plugin.version>1.23</jvnet-localizer-plugin.version>
     <forkCount>1</forkCount>
 
-    <nexus-platform-api.version>5.1.0-01</nexus-platform-api.version>
+    <nexus-platform-api.version>5.1.1</nexus-platform-api.version>
```
Comment (jmromeroes) [inline pom.xml:92]: We changed this for any specific reason? I see an issue brought by Eduard in the other [PR](https://github.com/sonatype/nexus-java-api/pull/278/files) which is similar so let's ask him what he means by that and see if it actually applies here as well.

File: src/main/java/org/sonatype/nexus/ci/iq/RemoteScanResult.groovy [hunk 2]
```diff
@@ -35,8 +47,16 @@ class RemoteScanResult
    */
   ScanResult copyToLocalScanResult() {
     def localFile = File.createTempFile('scan', '.xml.gz')
-    new FileOutputStream(localFile).withStream { filePath.copyTo(it) }
-    return new ScanResult(scan, localFile)
+    def additionalFiles = filePaths.stream().map {
+      File.createTempFile( it.baseName.split("\\.")[0],'.json.gz')
```
Comment (jmromeroes) [inline src/main/java/org/sonatype/nexus/ci/iq/RemoteScanResult.groovy:51]: I think it should always have `sast.json.gz` at the end but maybe @Qiming-C has more context about this.

Comment (oospinar) [reply src/main/java/org/sonatype/nexus/ci/iq/RemoteScanResult.groovy:51]: Yes, the name at the end is `sast.json.gz` but I made a generic way to get the name from the file path

File: src/main/java/org/sonatype/nexus/ci/iq/RemoteScanner.groovy [hunk 3]
```diff
@@ -98,7 +104,11 @@ class RemoteScanner
     def moduleIndices = getModuleIndices(workDirectory, moduleExcludes)
     def scanResult = iqClient.scan(appId, proprietaryConfig, advancedProperties, targets, moduleIndices, workDirectory,
         envVars, licensedFeatures)
-    return new RemoteScanResult(scanResult.scan, new FilePath(scanResult.scanFile))
+    return maybeGetSastJson(targets)
+        .map({ sastFile -> gzipFile(sastFile, workDirectory.toPath().toString()) })
+        .map ({sastFile -> new RemoteScanResult(scanResult.scan, new FilePath(scanResult.scanFile), List.of(new FilePath(sastFile)))})
```
Comment (jmromeroes) [inline src/main/java/org/sonatype/nexus/ci/iq/RemoteScanner.groovy:109]: The two maps have different formatting, this probably should be more like the one above.

File: src/test/java/org/sonatype/nexus/ci/iq/RemoteScannerTest.groovy [hunk 4]
```diff
@@ -243,4 +244,25 @@ class RemoteScannerTest
       assert arguments[0] == ['*.zip','*.tar']
     }
   }
+
+  def "create a sast file"() {
+    setup:
+      def remoteScanner = new RemoteScanner('appId', 'stageId', ['*.jar','!*.zip','*.war','!*.tar','!*.json'], [], workspace, proprietaryConfig, log,
+          "instance-id", null, null)
+      directoryScanner.getIncludedDirectories() >> []
+      directoryScanner.getIncludedFiles() >> ['scan.gz','sast.json']
```
Comment (jmromeroes) [inline src/test/java/org/sonatype/nexus/ci/iq/RemoteScannerTest.groovy:253]: Isn't it `scan.xml.gz`?

Comment (oospinar) [reply src/test/java/org/sonatype/nexus/ci/iq/RemoteScannerTest.groovy:253]: This is the file to compress.

**Comment by daniel-beck** [comment]
Closing this PR as the repository is being archived for https://github.com/jenkins-infra/helpdesk/issues/3742.

**Pull request closed without merging**